### PR TITLE
Sqlite: translators for string methods

### PR DESCRIPTION
**Pull request by SSkovboiSS**
Implements #10527

### CODE REVIEW

File: test/EFCore.Sqlite.FunctionalTests/Query/SimpleQuerySqliteTest.cs [hunk 1]
```diff
@@ -182,6 +182,36 @@ public override void Where_string_length()
 WHERE length(""c"".""City"") = 6");
         }
 
+        public override void Where_string_indexof()
+        {
+            base.Where_string_indexof();
```
Comment (smitpatel) [inline test/EFCore.Sqlite.FunctionalTests/Query/SimpleQuerySqliteTest.cs]: Add sql assertions for SqlServer also.

File: src/EFCore.Sqlite.Core/Query/ExpressionTranslators/Internal/SqliteStringIndexOfTranslator.cs [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+                var patternExpression = methodCallExpression.Arguments[0];
+
+                var zeroBasedCharIndexExpression = Expression.Subtract(
+                    new SqlFunctionExpression(
+                        "instr",
+                        typeof(int),
+                        new[] { methodCallExpression.Object, patternExpression }),
+                    Expression.Constant(1));
+
+                return patternExpression is ConstantExpression patternConstantExpression
```
Comment (smitpatel) [inline src/EFCore.Sqlite.Core/Query/ExpressionTranslators/Internal/SqliteStringIndexOfTranslator.cs]: You don't need special processing for empty string. Same formula returns correct result. No need to over-optimize.

File: src/EFCore.Sqlite.Core/Query/ExpressionTranslators/Internal/SqliteStringIndexOfTranslator.cs [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+
+        /// <summary>
+        ///     This API supports the Entity Framework Core infrastructure and is not intended to be used
+        ///     directly from your code. This API may change or be removed in future releases.
+        /// </summary>
+        public virtual Expression Translate(MethodCallExpression methodCallExpression)
+        {
+            if (Equals(methodCallExpression.Method, _methodInfo))
+            {
+                var patternExpression = methodCallExpression.Arguments[0];
```
Comment (smitpatel) [inline src/EFCore.Sqlite.Core/Query/ExpressionTranslators/Internal/SqliteStringIndexOfTranslator.cs]: Inline variable.

File: src/EFCore.Sqlite.Core/Query/ExpressionTranslators/Internal/SqliteStringSubstringTranslator.cs [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+        /// </summary>
+        public virtual Expression Translate(MethodCallExpression methodCallExpression)
+            => _methodInfo.Equals(methodCallExpression.Method)
+                ? new SqlFunctionExpression(
+                    "substr",
+                    methodCallExpression.Type,
+                    new[]
+                    {
+                        methodCallExpression.Object,
+                        methodCallExpression.Arguments[0].NodeType == ExpressionType.Constant
```
Comment (smitpatel) [inline src/EFCore.Sqlite.Core/Query/ExpressionTranslators/Internal/SqliteStringSubstringTranslator.cs]: Use pattern match like
```
methodCallExpression.Arguments[0] is constantExpression
&& constantExpression.Value is int value
? Expression.Constant(value + 1)
: Expression.Add(....)
```

File: src/EFCore.Specification.Tests/Query/SimpleQueryTestBase.Where.cs [hunk 5]
```diff
@@ -610,6 +610,31 @@ public virtual void Where_string_length()
                 entryCount: 20);
         }
 
+        [ConditionalFact]
```
Comment (smitpatel) [inline src/EFCore.Specification.Tests/Query/SimpleQueryTestBase.Where.cs]: Add test for IndexOf with empty string
Replace with empty string replacement to check erase
SubString with startIndex = 0 and/or length = 0

File: test/EFCore.Sqlite.FunctionalTests/Query/SimpleQuerySqliteTest.cs (hunk 1 above)
Comment (SSkovboiSS) [reply test/EFCore.Sqlite.FunctionalTests/Query/SimpleQuerySqliteTest.cs]: What is the relation of SqlServer to Sqlite test?
I am not sure if I understand what are you asking for.

Comment (smitpatel) [reply test/EFCore.Sqlite.FunctionalTests/Query/SimpleQuerySqliteTest.cs]: Override this methods in SimpleQuerySqlServerTest class and assert SQL generated.
All functional tests are run against all the providers. So even you are making change in SQLite only, the test will run on SqlServer too. Since SqlServer provides most functionality, we always assert SQL for all sync tests in SqlServer.

**Comment by SSkovboiSS** [comment]
> Can you translate this into SqlServer?

SqlServer translators for Substring and Replace already exists. I will add translator for IndexOf. Should I append commit for it to this PR?

Also it seems like `SqlServerStringReplaceTranslator` need more tests. I can add them as well.

**Comment by smitpatel** [comment]
@SSkovboiSS - Yes adding into this PR is good. Also Substring in SQL Server can use pattern matching like above.

More tests are always welcome.

File: src/EFCore.Specification.Tests/Query/SimpleQueryTestBase.Functions.cs [hunk 6]
```diff
@@ -816,6 +816,51 @@ public virtual void Convert_ToString()
...
+                    context.Set<Customer>().OrderBy(c => c.CustomerID).Select(c => c.ContactName.IndexOf(string.Empty)).First());
+            }
+        }
+
+        [ConditionalFact]
+        public virtual void Replace_with_emptystring()
+        {
+            using (var context = CreateContext())
+            {
+                Assert.Equal(
```
Comment (smitpatel) [inline src/EFCore.Specification.Tests/Query/SimpleQueryTestBase.Functions.cs]: Any reason this are not using AssertQuery syntax?
I think the function for single result would be AssertSingleResult.
Also use a predicate on Customer like `c => c.CustomerID == "ALFKI"` to reduce the data fetched from server.

File: test/EFCore.SqlServer.FunctionalTests/Query/SimpleQuerySqlServerTest.Functions.cs [hunk 7]
```diff
@@ -1038,6 +1038,54 @@ public override void Convert_ToString()
...
         }
 
+        public override void Indexof_with_emptystring()
+        {
+            base.Indexof_with_emptystring();
+
+            AssertSql(
+                @"@__Empty_0='' (Size = 4000)
+
+SELECT TOP(1) CASE
```
Comment (smitpatel) [inline test/EFCore.SqlServer.FunctionalTests/Query/SimpleQuerySqlServerTest.Functions.cs]: Even though the arg - string.Empty is inlined, it still created parameter for. If the constant of string empty is never going to appear, would it be necessary to optimize the constant case in translator?
I will run some tests tomorrow to check out what is happening. I am commenting since i found it interesting.

Comment (SSkovboiSS) [reply test/EFCore.SqlServer.FunctionalTests/Query/SimpleQuerySqlServerTest.Functions.cs]: I replaced string.Empty with a "" in a test and parameter has gone but it seems like client evalueted now, because it produced the following SQL:
SELECT TOP(1) [c].[ContactName]
FROM [Customers] AS [c]
ORDER BY [c].[CustomerID]

File: src/EFCore.Specification.Tests/Query/SimpleQueryTestBase.Functions.cs (hunk 6 above)
Comment (SSkovboiSS) [reply src/EFCore.Specification.Tests/Query/SimpleQueryTestBase.Functions.cs]: Well there are other tests (e.g. for Substring) in SimpleQuerySqlServerTest.Functions that follows same pattern, so I just borrowed it from there:)

Comment (SSkovboiSS) [reply src/EFCore.Specification.Tests/Query/SimpleQueryTestBase.Functions.cs]: Actually I can just refactor all of them - there is no problem about that.

Also use a predicate on Customer like c => c.CustomerID == "ALFKI" to reduce the data fetched from server.
Does this really necessary, considering that I use First?

Comment (smitpatel) [reply src/EFCore.Specification.Tests/Query/SimpleQueryTestBase.Functions.cs]: Its a minor thing. Few months ago, i wouldn't have said so. But while working on writing a provider for EF, the test time issues i ran into are glaring. In brief, We translate `First` to server so it is not going to fetch a lot of data but a provider which cannot translate `First` to server have to fetch whole table. Unless test really requires lack of predicate, it should restrict result set. Or you can use table like Employees which has only 9 rows.

File: test/EFCore.SqlServer.FunctionalTests/Query/SimpleQuerySqlServerTest.Functions.cs (hunk 7 above)
Comment (smitpatel) [reply test/EFCore.SqlServer.FunctionalTests/Query/SimpleQuerySqlServerTest.Functions.cs]: @SSkovboiSS - I filed #10547 That is the cause why string.empty created parameter. (so we need to use "") for now.
It client eval'ed because of some optimization we do in projection when it translates to constant. Filed #10548

Can you revert this optimization for now?
Just return Conditional Expression always. Likewise for Sqlite

File: src/EFCore.SqlServer/Query/ExpressionTranslators/Internal/SqlServerStringIndexOfTranslator.cs [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+                var patternExpression = methodCallExpression.Arguments[0];
+
+                var charIndexExpression = Expression.Subtract(
+                    new SqlFunctionExpression(
+                        "CHARINDEX",
+                        typeof(int),
+                        new[] { patternExpression, methodCallExpression.Object }),
+                    Expression.Constant(1));
+
+                return patternExpression is ConstantExpression patternConstantExpression
```
Comment (smitpatel) [inline src/EFCore.SqlServer/Query/ExpressionTranslators/Internal/SqlServerStringIndexOfTranslator.cs]: Can you remove the optimization return the constant?
If the patternExpression is constant and not null or empty then return charIndexExpression. In all other cases, conditional block. Also check for IsNullOrEmpty because null can appear.

File: src/EFCore.SqlServer/Query/ExpressionTranslators/Internal/SqlServerStringSubstringTranslator.cs [hunk 9]
```diff
@@ -29,12 +29,12 @@ public virtual Expression Translate(MethodCallExpression methodCallExpression)
...
                         methodCallExpression.Object,
                         // Accommodate for SQL Server assumption of 1-based string indexes
-                        methodCallExpression.Arguments[0].NodeType == ExpressionType.Constant
-                            ? (Expression)Expression.Constant(
-                                (int)((ConstantExpression)methodCallExpression.Arguments[0]).Value + 1)
-                            : Expression.Add(
-                                methodCallExpression.Arguments[0],
-                                Expression.Constant(1)),
+                        methodCallExpression.Arguments[0] is ConstantExpression constantExpression
+                            && constantExpression.Value is int value
```
Comment (smitpatel) [inline src/EFCore.SqlServer/Query/ExpressionTranslators/Internal/SqlServerStringSubstringTranslator.cs]: Sorry for my earlier comment But since the first arg is always int type, you can remove the check `constantExpression.Value is int` and instead do hard cast like `(int)constantExpression.Value`
Same in SQLite.

File: test/EFCore.Sqlite.FunctionalTests/Query/SimpleQuerySqliteTest.cs (hunk 1 above)
Comment (smitpatel) [reply test/EFCore.Sqlite.FunctionalTests/Query/SimpleQuerySqliteTest.cs]: #Resolved

File: src/EFCore.SqlServer/Query/ExpressionTranslators/Internal/SqlServerStringIndexOfTranslator.cs (hunk 8 above)
Comment (SSkovboiSS) [reply src/EFCore.SqlServer/Query/ExpressionTranslators/Internal/SqlServerStringIndexOfTranslator.cs]: Like that?

return patternExpression is ConstantExpression constantExpression
                    && !string.IsNullOrEmpty((string)constantExpression.Value)
                        ? (Expression)charIndexExpression
                        : Expression.Condition(
                            Expression.Equal(patternExpression, Expression.Constant(string.Empty)),
                            Expression.Constant(0),
                            charIndexExpression);

It produces SQL:
SELECT CASE
    WHEN N'' = N''
    THEN 0 ELSE CHARINDEX(@__Empty_0, [c].[ContactName]) - 1
END
FROM [Customers] AS [c]
WHERE [c].[CustomerID] = N'ALFKI'

Comment (smitpatel) [reply src/EFCore.SqlServer/Query/ExpressionTranslators/Internal/SqlServerStringIndexOfTranslator.cs]: Yup. That sounds correct.

**Comment by smitpatel** [comment]
Merged via 8886d1ca746a6730d4ec01a00e79450620e915f4

@SSkovboiSS  Thanks for contribution.

File: src/EFCore.SqlServer/Query/ExpressionTranslators/Internal/SqlServerStringIndexOfTranslator.cs [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+                var patternExpression = methodCallExpression.Arguments[0];
+
+                var charIndexExpression = Expression.Subtract(
+                    new SqlFunctionExpression(
+                        "CHARINDEX",
+                        typeof(int),
+                        new[] { patternExpression, methodCallExpression.Object }),
+                    Expression.Constant(1));
+
+                return patternExpression is ConstantExpression constantExpression
```
Comment (smitpatel) [inline src/EFCore.SqlServer/Query/ExpressionTranslators/Internal/SqlServerStringIndexOfTranslator.cs]: Missed picking up this bit in last commit. Merged via 778eeb31f5b79d1d47231fc34479d459fc4e7248
